### ISSUE TITLE
Missing items after direct-read

### ISSUE THREAD

**Issue by VitaliStupin** [rwynn/monstache#97]
I found a problem that when I transfer full MongoDB collection (100k+ items) with "direct-read-namespaces" option then a small amount of items is missing in Elasticsearch. I also found that these missing items usually come in one or couple of batches of items ordered by order of insertion into MongoDB. Therefore I assume that some of the batches received from MongoDB just get lost in the process. I also was trying the verbose option and these missing items were not found in the verbose log, so they disappear before that logging is done.

To reproduce the problem I am using:
Ubuntu 16.04.4 LTS
MongoDB version 4.0.0 (also reproducible with v3.4.15)
Elasticsearch version 6.3.1
Monstache version 4.9.0

MongoDB and Elasticsearch are used with default configurations (only enabling replication in MongoDB). Repositories:
```bash
$ cat /etc/apt/sources.list.d/elastic-6.x.list
deb https://artifacts.elastic.co/packages/6.x/apt stable main
$ cat /etc/apt/sources.list.d/mongodb-org-4.0.list
deb [ arch=amd64,arm64 ] https://repo.mongodb.org/apt/ubuntu xenial/mongodb-org/4.0 multiverse
```

Helper scripts:
```
$ cat add_items_to_mongo.py
#!/usr/bin/python3
from pymongo import MongoClient
client = MongoClient()
collection = client.test_database.test_collection
for i in range (0, 1000):
    posts = []
    for j in range (0, 1000):
        posts.append({'id': i * 1000 + j, 'data': '12345678901234567890'})
    collection.insert_many(posts)

$ cat ids.py
#!/usr/bin/python3
from elasticsearch import Elasticsearch
es = Elasticsearch()
page = es.search(index='test_database.test_collection', size=5000, scroll='2m')
sid = page['_scroll_id']
scroll_size = len(page['hits']['hits'])
while scroll_size > 0:
    for hit in page['hits']['hits']:
        print(hit['_source']['id'])
    page = es.scroll(scroll_id=sid, scroll='2m')
    sid = page['_scroll_id']
    scroll_size = len(page['hits']['hits'])
```

Monstache configuration:
```
$ cat config.toml
mongo-url = "mongodb://localhost:27017"
elasticsearch-urls = ["http://localhost:9200"]
namespace-regex = '^test_database\.test_collection$'
direct-read-namespaces = ["test_database.test_collection"]
gzip = true
stats = true
index-stats = false
index-oplog-time = true
dropped-collections = false
dropped-databases = false
replay = false
resume = false
verbose = false
exit-after-direct-reads = true
enable-http-server = false
elasticsearch-max-conns = 1
```

Inserting data into MongoDB:
```
$ python3 add_items_to_mongo.py

$ mongo test_database --eval "db.test_collection.count()"
MongoDB shell version v4.0.0
connecting to: mongodb://127.0.0.1:27017/test_database
MongoDB server version: 4.0.0
1000000
```

Running Monstache:
```
$ ./monstache -f config.toml
INFO 2018/08/15 16:31:49 Successfully connected to MongoDB version 4.0.0
INFO 2018/08/15 16:31:49 Successfully connected to Elasticsearch version 6.3.1
INFO 2018/08/15 16:31:49 Entering event loop
STATS 2018/08/15 16:32:20 {"Flushed":6,"Committed":10,"Indexed":246989,"Created":0,"Updated":0,"Deleted":0,"Succeeded":246989,"Failed":0,"Workers":[{"Queued":0,"LastDuration":1833000000}]}
STATS 2018/08/15 16:32:51 {"Flushed":12,"Committed":21,"Indexed":512363,"Created":0,"Updated":0,"Deleted":0,"Succeeded":512363,"Failed":0,"Workers":[{"Queued":0,"LastDuration":1713000000}]}
STATS 2018/08/15 16:33:21 {"Flushed":18,"Committed":33,"Indexed":756019,"Created":0,"Updated":0,"Deleted":0,"Succeeded":756019,"Failed":0,"Workers":[{"Queued":0,"LastDuration":1347000000}]}
STATS 2018/08/15 16:33:50 {"Flushed":24,"Committed":43,"Indexed":996763,"Created":0,"Updated":0,"Deleted":0,"Succeeded":996763,"Failed":0,"Workers":[{"Queued":0,"LastDuration":1044000000}]}
INFO 2018/08/15 16:33:50 Shutting down
```

Checking the amount of data in Elasticsearch:
```
$ curl http://127.0.0.1:9200/_cat/indices/test_database.test_collection?v
health status index                         uuid                   pri rep docs.count docs.deleted store.size pri.store.size
yellow open   test_database.test_collection czteAsvnTKSpGaD2TYZuzQ   5   1     998989            0     41.4mb         41.4mb
```

As it can be seen some documents are missing in Elasticsearch (less then 1M of items).

Now to check which items are missing lets create a list of numbers from 0 to 999999
```
$ python3 -c "for i in range (0, 1000000): print(i)" > all_ids.txt
```

Then load id values from Elasticsearch and compare with reference list:
```
$ python3 ids.py | sort -n > ids.txt

$ diff ids.txt all_ids.txt
```

Diff shows that documents from 998989 to 999999 are missing.

If I repeat the test by deleting index in elastic search (`curl -X DELETE "localhost:9200/test_database.test_collection"`) and then repeat steps starting from running Monstache. I get different result like:

798978 ... 799999 is missing
499287 ... 499999 is missing
etc...

**Comment by rwynn** [comment]
Can you take out exit-after-direct-reads and see if they all make it eventually.

**Comment by VitaliStupin** [comment]
I have tested now couple of times without the exit-after-direct-reads option and you are right, it does work better this way, and all the documents get transfrered.

Does that mean that separate threads that fetch data from MongoDB are lacking some lock mechanism and the main program exits before all threads are finished transferring the data?

That makes me think that one other problem I am experiencing is also related to that.

Namely we have one big collection which fails to transfer. In the monstache log we see:
```
ERROR 2018/08/16 11:31:58 Error reading segment of collection db_name.collection_name. Will retry segment.: read tcp 10.0.xxx.xxx:38182->10.0.xxx.yyy:27017: i/o timeout
```
While in the same time in the MongoDB log we see:
```
2018-08-16T08:38:17.269+0300 I COMMAND  [conn912587] killcursors db_name.collection_name exception: Cannot kill pinned cursor: 185143284434 code:16089 numYields:0 locks:{ Global: { acquireCount: { r: 2 } }, Database: { acquireCount: { r: 1 } }, Collection: { acquireCount: { r: 1 } } } 0ms
```

Do I understand correctly that:
https://github.com/rwynn/gtm/blob/master/gtm.go#L978
iter.Close() is sending killcursors to MongoDB before all the threads are finished transferring segment?

**Comment by rwynn** [comment]
All the threads that fetch data put each record onto a channel. When each thread finishes it signals a wait group. 

Monstache waits on the wait group in the case of exit-after.  When this lock gets released it means that all the readers are finished writing to a channel.  Monstache does a flush to Elasticsearch at this point. 

But there could still be work that has not been read off that channel and queued for delivery to Elasticsearch. 

I will have to investigate a better way to determine if all these various queues are empty to make sure exit-after does not exit too soon. 

It may help if you reduce the gtm-settings -> channel-size setting to 1. 

[gtm-settings]
channel-size = 1

**Comment by rwynn** [comment]
Think I have a fix for this. I’ll update this issue when I get a chance to test and commit.  Might be a little while due to some other priorities I have.

**Comment by rwynn** [comment]
@VitaliStupin this should be fixed now in the master branch.  You would need to build it manually unit I get a chance to make a formal release

**Comment by VitaliStupin** [comment]
Thank you! Confirming that direct-read-namespaces is now working properly with "exit-after-direct-reads = true".

I will need some time to also test the second problem with "Cannot kill pinned cursor" errors.

**Comment by VitaliStupin** [comment]
It seems that the i/o timeout problem with default channel-size ("Cannot kill pinned cursor" in mongo log) is still present. So it is a different issue. But this time segment repeating did not last infinitely, and the process finished successfully. So the fix had positive influence, but some different problem may still exist.

**Comment by rwynn** [comment]
@VitaliStupin from what I've read applying the socket timeout setting like in this commit might fix it.

https://github.com/mongodb/mongo-tools/commit/3aec8b5e7fb1a7473fd3b705f0220a721745ff77

**Comment by rwynn** [comment]
Are you able to try setting the 3 timeouts under mongo-dial-settings and mongo-session-settings to something like 600 each. That’s seconds.  

I have an upcoming change to make more timeouts configurable. But until then you might try the upping existing options. 

I think the client is forcing the socket to close after not getting timely response and trying to kill the cursor but MongoDB is still working on it.

**Comment by VitaliStupin** [comment]
I have actually tried increasing timeouts previously with:
[mongo-session-settings]
socket-timeout = 300
sync-timeout = 300

And that did not give any results, I did not try to increase the connection timeout because mongo gives error about closing existing connection, therefore connection was created and used.

I also checked in mongodb log that all queries were no longer than 1 second, so there should not be issues with too long queries.

**Comment by rwynn** [comment]
Have you tried the latest monstache?  I surfaced more timeouts from the mgo driver and upped the default values. 

I also noticed that the 10s initial connect timeout was being used as the read and write timeout if either of those was not explicitly defined. So I’ve now explicitly defined those 2.

**Comment by rwynn** [comment]
@VitaliStupin I will close this issue about the missing items.  Please open a new issue if you still have i/o timeout with 4.10.1.
